src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -0,0 +1,164 @@
+                        ? resultUsedLength + 1
+                        : resultUsedLength;
+
+                    if (destinationOffset < result.Length)
+                    {
+                        Span<char> destination = result.Slice(destinationOffset);
+
+                        written = _lowercase
+                            ? word.ToLowerInvariant(destination)
+                            : word.ToUpperInvariant(destination);
+
+                        if (written > 0)
+                        {
+                            break;
+                        }
+                    }
+
+                    ExpandBuffer(ref result);
+                }
+
+                if (resultUsedLength != 0)
+                {
+                    result[resultUsedLength] = _separator;
+                    resultUsedLength += 1;
+                }
+
+                resultUsedLength += written;
+            }
+
+            int first = 0;
+            ReadOnlySpan<char> chars = name.AsSpan();
+            CharCategory previousCategory = CharCategory.Boundary;
+
+            for (int index = 0; index < chars.Length; index++)
+            {
+                char current = chars[index];
+                UnicodeCategory currentCategoryUnicode = char.GetUnicodeCategory(current);
+
+                if (currentCategoryUnicode == UnicodeCategory.SpaceSeparator ||
+                    currentCategoryUnicode >= UnicodeCategory.ConnectorPunctuation &&
+                    currentCategoryUnicode <= UnicodeCategory.OtherPunctuation)
+                {
+                    WriteWord(chars.Slice(first, index - first), ref result);
+
+                    previousCategory = CharCategory.Boundary;
+                    first = index + 1;
+
+                    continue;
+                }
+
+                if (index + 1 < chars.Length)
+                {
+                    char next = chars[index + 1];
+                    CharCategory currentCategory = currentCategoryUnicode switch
+                    {
+                        UnicodeCategory.LowercaseLetter => CharCategory.Lowercase,
+                        UnicodeCategory.UppercaseLetter => CharCategory.Uppercase,
+                        _ => previousCategory
+                    };
+
+                    if (currentCategory == CharCategory.Lowercase && char.IsUpper(next) ||
+                        next == '_')
+                    {
+                        WriteWord(chars.Slice(first, index - first + 1), ref result);
+
+                        previousCategory = CharCategory.Boundary;
+                        first = index + 1;
+
+                        continue;
+                    }
+
+                    if (previousCategory == CharCategory.Uppercase &&
+                        currentCategoryUnicode == UnicodeCategory.UppercaseLetter &&
+                        char.IsLower(next))
+                    {
+                        WriteWord(chars.Slice(first, index - first), ref result);
```

Comment:
I'm curious - should we be using identical rules to CamelCase except also adding `-` or `_`?

Reply:
Sure, I can do that, but knowing about "compatibility concerns" I expect to hear "no" from other reviewers (:

Reply:
And actually, what you asked is how naming policies work in other projects and platforms. They always split the input and then transform each word and concatenate them back.

Reply:
Please file an issue to consider joining this code with CamelCasing and I think we're ready to merge after that